keps/sig-node/4960-container-stop-signals/README.md

```diff
@@ -0,0 +1,471 @@
+
+### Graduation Criteria
+
+#### Alpha
+
+- Feature implemented behind a feature flag
+- CRI API implementation completed in containerd marked as experimental
+- CRI API implementation completed for CRI-O
+- Initial e2e tests completed and enabled, testing the feature against containerd
+- Unit tests for validation, e2e tests for version skew
+
+#### Beta
+
+- Add support for Windows
+- Gather feedback from developers and surveys
+- e2e tests for CRI-O
+
+#### GA
+
+- Both containerd and CRI-O having a GA release with containerStopSignal parameter implemented for `stopContainer` method.
+- More rigorous forms of testing—e.g., downgrade tests and scalability tests
+- Allowing time for user feedback
+
+### Upgrade / Downgrade Strategy
+
+#### Upgrade
+
+When upgrading to a new Kubernetes version which supports Container Stop Signals, users can enable the feature gate and start using the feature. If the user is running an older version of the container runtime, the feature will be gracefully degraded as mentioned [here](https://www.kubernetes.dev/docs/code/cri-api-version-skew-policy/#version-skew-policy-for-cri-api) in the CRI API version skew doc. In this case the user will be able to set a StopSignal lifecycle hook in the Container spec, but the kubelet will not pass this value to the container runtime when calling the `runtimeService.stopContainer` method. The container status would also not have stop signal since the container runtime is not updated to return the effective stop signal extracted from the image.
+
+#### Downgrade
+
+If the kube-apiserver or the kubelet's version is downgraded, you will no longer be able to create or update container specs to include the StopSignal lifecycle hook. Existing containers which have the field set would not be cleared. If you're running a version of the kubelet which doesn't support ContainerStopSignals, the CRI API wouldn't pass the stop signal to the runtime as part of ContainerConfig. Even if the container runtime is on a newer version supporting stop signal, it would handle this and default to the stop signal defined in the image or to SIGTERM.
+
+### Version Skew Strategy
+
+Both kubelet and kube-apiserver will need to enable the feature gate for the full featureset to be present and working. If both components disable the feature gate, this feature will be completely unavailable.
+
+If only the kube-apiserver enables this feature, validation will pass, but kubelet won't understand the new lifecycle hook and will not add the stop signal when creating the ContainerConfig.
+
+If only the kubelet has enabled this feature, you won't be able to create a Pod which has a StopSignal lifecycle hook via the apiserver and hence the feature won't be usable even if the kubelet supports it. `containerSpec.Lifecycle.StopSignal` can be an empty value and kubelet functions as if no custom stop signal has been set for any container.
```

Comment:
What if the user create a static pod with this feature?

Reply:
If the feature is disabled in the kubelet and if we create a static pod, the StopSignal lifecycle hook would be dropped by the kubelet. In the case where the feature gate is disabled in the apiserver but enabled on the kubelet, the kubelet would pass the stop signal to the container runtime and it would work, but the kube-apiserver wouldn't show the pod as having a custom stop signal. This is the same case that'd happen when we create a regular pod with a stop signal and later disable the feature gate in kube-apiserver. The feature would work as long as the feature gate is enabled in the kubelet. 

What do you think?

Reply:
Got it. 
/lgtm